doc/man/lfescript.1
.\" Automatically generated by Pandoc 2.11.2
.\"
.TH "lfescript" "1" "2013-2016" "" ""
.hy
.SH NAME
.PP
lfescript - Lisp Flavoured Erlang (LFE) scripting support
.SH SYNOPSIS
.PP
lfescript provides support for running short LFE programs without having
to compile them first and an easy way to retrieve the command line
arguments.
.SH EXPORTS
.PP
\f[B]script-name script-arg1 script-arg2 \&...\f[R]
.PP
\f[B]lfescript lfescript-flags script-name script-arg1 script-arg2
\&...\f[R]
.PP
lfescript runs a script written in LFE.
.PP
\f[B]lfescript:script_name() -> File\f[R]
.PP
Types:
.IP
.nf
\f[C]
File = filename()
\f[R]
.fi
.PP
The \f[C]script_name/0\f[R] function returns the name of the lfescript
being executed.
If the function is invoked outside the context of an lfescript, the
behavior is undefined.
.SH EXAMPLE
.PP
Here follows an example script.
.IP
.nf
\f[C]
$ cat factorial
#! /usr/bin/env lfescript
;; -*- mode: lfe -*-
;;! -smp enable -sname factorial -mnesia debug verbose

(defun main
  ([(list string)]
   (try
       (let* ((n (list_to_integer string))
              (f (fac n)))
         (lfe_io:format \[dq]factorial \[ti]w = \[ti]w\[rs]n\[dq] (list n f)))
     (catch
       ((tuple _ _ _) (usage)))))
  ([_] (usage)))

(defun fac
  ([0] 1)
  ([n] (* n (fac (- n 1)))))

(defun usage ()
  (lfe_io:format \[dq]usage: factorial integer\[rs]n\[dq] ()))
\f[R]
.fi
.PP
The header of the LFE script is different from a normal LFE module.
The first line is an interpreter line which invokes lfescript if the
script is run as in the first command line above.
If lfescript is explicitly invoked in the second command line above then
this header line will be ignored.
On the second or third line it is possible to give arguments to the
emulator with the syntax:
.IP
.nf
\f[C]
;;! -smp enable -sname factorial -mnesia debug verbose
\f[R]
.fi
.PP
In the example the second line is an optional directive to Emacs which
causes it to enter LFE mode when editing the script file.
.PP
The rest of the file contains LFE source code.
It must always contain the function \f[C]main/1\f[R].
When the script is run this function will be called with a list of
strings representing the arguments with which the script was called.
It is possible to define, include and use macros in the source code.
.PP
The source code is checked and warnings and errors will be printed.
If there are errors the script will not run and it will terminate with
exit status 127.
Otherwise the code will be interpreted.
If the function \f[C]main/1\f[R] returns successfully then the exit
status for the script will be 0 but if an exception is raised then exit
status will be 127.
.SH OPTIONS
.PP
The following option is accepted by lfescript
.IP \[bu] 2
\f[C]-s\f[R] - Only perform a syntactic and semantic check of the script
file.
Warnings and errors (if any) are written to the standard output, but the
script will not be run.
The exit status will be 0 if there were no errors, and 127 otherwise.
.PP
Unrecognised options are ignored.
.SH ENVIRONMENT VARIABLES
.PP
\f[B]LFESCRIPT_EMULATOR\f[R]
.PP
The command used to start the emulator.
Default is `erl'.
This can be useful for passing arguments into the emulator, for example
.IP
.nf
\f[C]
LFESCRIPT_EMULATOR=\[dq]erl -pa sune\[dq]
\f[R]
.fi
.PP
will add the directory sune to the code path.
.SH SEE ALSO
.PP
\f[B]lfe(1)\f[R], \f[B]lfe_guide(7)\f[R]
.SH AUTHORS
Robert Virding.


src/ChangeLog
2013-02-12  Robert Virding  <rv@renat.local>

    * <All files>: Changed Copyright notice and now use Apache License
    V2.0.

2011-08-19  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (pass_expand_expr): Only try to expand lists.

2011-07-09  Robert Virding  <rv@renat.local>

    * lfe_trans.erl (from_bitseg): Fix handling of binaries.

2011-07-02  Robert Virding  <rv@renat.local>

    * lfe_eval.erl (match_binary): Fix handling of binary and pattern
    vars.

    * lfe_lint.erl (pat_binary): Fix handling of binary and pattern
    vars.

2011-06-29  Robert Virding  <rv@renat.local>

    * lfe_lint.erl, lfe_codegen.erl, lfe_eval.erl, lfe_trans.erl: Now
    use lfe_bits to parse binaries.

2011-06-26  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add match-spec and qlc macros.

2011-06-15  Robert Virding  <rv@renat.local>

    * lfe_shell.erl (macroexpand, macroexpand-1, macroexpand-all): No
    longer shell functions.

    * lfe_lib.erl (del_vbinding): New function.

2011-06-14  Robert Virding  <rv@renat.local>

    * lfe_io_pretty.erl (indent_type): Add funcall.

2011-06-08  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add match-spec macro.

2011-06-05  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add simple include-lib macro.

2011-05-30  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_defrecord): Add ets/mnesia match pattern
    macro emp-XXX.

    * lfe_eval.erl, lfe_shell.erl, lfe_macro.erl: Restructure
    lfe_macro.erl and redefine interface.

2011-04-24  Robert Virding  <rv@renat.local>

    * lfe_codegen.erl (comp_gtest): Removed unnecessary special cases,
    optimizer fixes this for us.

2011-04-21  Robert Virding  <rv@renat.local>

    * lfe_lint.erl: Remove unused erroneous lfe_lib imports.

2011-04-19  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Arithmetic and comparison functions
    take multiple arguments.

2011-04-12  Robert Virding  <rv@renat.local>

    * lfe_comp.erl (erl_comp_opts, do_ok_return): Handle
    warnings_as_errors ourselves in same way as vanilla compiler.

2011-04-08  Robert Virding  <rv@renat.local>

    * lfe_comp.erl (when_opt): Change argument order.

2011-03-16  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (expand_append): Expand ++ in such a way as to
    allow use in patterns.
    (expand_defmacro): Allow (defmacro name arg ...).

2010-12-31  Robert Virding  <rv@renat.local>

    * lfe_io_format.erl (control, newline): Handle ~n correctly.

    * lfe_io_pretty.erl (print1): Prettyprint one element lists.

2010-11-25  Robert Virding  <rv@renat>

    * lfe_lint.erl (pat_bitsegs): Use check_foldl instead.
    (gexpr_bitsegs): Use check_foreach instead.

2010-11-04  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Extend ? macro to handle optional
    timeout and default value.

2010-11-01  Robert Virding  <rv@renat>

    * lfe_scan.xrl: Check creating atoms/symbols and code cleanup.

2010-10-27  Robert Virding  <rv@renat.local>

    * lfe_shell.erl (prompt): Prompt adds node name when alive.

2010-09-06  Robert Virding  <rv@stanislaw.local>

    * lfe_macro.erl (def_record, def_rec_fields): Record macros follow
    new pattern code.
    (expand_defmacro): Add explicit list constructor for args.
    (c_l_tq): Add explicit cons constructor.

    * lfe_codegen.erl (comp_pat, pat_list): Add code to generate
    pattern code for explicit list constructors.

    * lfe_eval.erl (match, match_list): Add code to match explicit
    list constructors.

    * lfe_lint.erl (check_pat, pat_list, check_alias): Add code to
    test for explicit list constructors.

2010-08-25  Robert Virding  <rv@stanislaw.local>

    * lfe_macro.erl (c_tq, c_l_tq, c_b_tq): Add guards to generators
    in comprehensions.

    * lfe_lint.erl (format_error, check_gbody, check_gexpr)
    (gexpr_bitsegs): Better error checking in guards.

2010-07-24  Robert Virding  <rv@host-90-237-128-188.mobileonline.telia.com>

    * lfe_shell.erl (macroexpand, macroexpand_1): Return evaluated
    argument if no expansion.

2010-07-12  Robert Virding  <rv@host-90-233-145-108.mobileonline.telia.com>

    * lfe_scan.xrl: Have separate cases for integer, float and symbol
    atoms, catch errors better.

2010-06-07  Robert Virding  <rv@stanislaw.local>

    * lfe_lint.erl (expr_bitseg, gexpr_bitseg, pat_bitseg): Allow
    strings (integer lists).

    * lfe_codegen.erl (parse_bitsegs): Allow strings (integer lists).

    * lfe_eval.erl (parse_bitsegs): Allow strings (integer lists).

    * lfe_eval.lfe (eval-gbinary): Evaluate binaries in guards.

    * lfe_eval.erl (eval_gbinary): Evaluate binaries in guards.

2010-06-03  Robert Virding  <rv@stanislaw.local>

    * All files: Update copyright to 2010.

2010-04-30  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.lfe (apply-lambda): New name and more logical argument
    order for eval-lambda/4.
    (apply-match-clauses): New name and more logical argument
    order for eval-match-clauses/3.

2010-02-24  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.erl (apply_lambda): New name and more logical argument
    order for eval_lambda/4.
    (apply_match_clauses): New name and more logical argument
    order for eval_match_clauses/3.

    * lfe_macro.erl (expand_clause, expand_ml_clause, exp_predef)
    (c_tq): Guards are now a sequence of tests (when t ...).

2010-02-20  Robert Virding  <rv@stanislaw.local>

    * lfe_io.erl (scan_and_parse): Use new re-entrant parser.

    * lfe_parse.erl: Write to become re-entrant using LL(1) grammar.

2010-02-16  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.lfe (eval-guard, match-when, eval-let): Guards are now
    a sequence of tests (when t ...).

    * lfe_eval.erl (eval_guard, match_when, eval_let): Guards are now
    a sequence of tests (when t ...).

    * lfe_codegen.erl (comp_guard, comp_gtest, comp_clause_body)

src/lfescript.erl
%% Copyright (c) 2008-2020 Robert Virding
%%
%% Licensed under the Apache License, Version 2.0 (the "License");
%% you may not use this file except in compliance with the License.
%% You may obtain a copy of the License at
%%
%%     http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing, software
%% distributed under the License is distributed on an "AS IS" BASIS,
%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%% See the License for the specific language governing permissions and
%% limitations under the License.

%%% File    : lfescript.erl
%%% Author  : Robert Virding
%%% Purpose : Lisp Flavoured Erlang scripting interface.

-module(lfescript).

%% External API.
-export([script_name/0]).

%% Internal API.
-export([start/0,start/1]).
-export([run/1,run/2]).

-include("lfe.hrl").

%% External API.

script_name() ->
    [Sname|_] = init:get_plain_arguments(),
    Sname.

%% Internal API.

-define(OK_STATUS, 0).
-define(ERROR_STATUS, 127).

-spec start() -> no_return().
-spec start(_Options) -> no_return().
-spec run(_CmdLine) -> no_return().
-spec run(_CmdLine, _Options) -> no_return().
%%  Evaluate the LFE script. All errors which are caught here are
%%  internal errors. Start gets its arguments from the command line
%%  while run gets them as an argument.

start() -> start([]).

start(Lopts) ->
    run(init:get_plain_arguments(), Lopts).

run(CmdLine) -> run(CmdLine, []).

run([File|Args], Lopts) ->
    try
        process_flag(trap_exit, false),
        parse_check_run(File, Args, Lopts)
    catch
        %% Catch program errors.
        throw:Str ->
            lfe_io:format("lfescript: ~s\n", [Str]),
            halt(?ERROR_STATUS);
        ?CATCH(_, Reason, Stack)
            lfe_io:format("lfescript: Internal error: ~p\n", [Reason]),
            lfe_io:format("~p\n", [Stack]),
            halt(?ERROR_STATUS)
    end;
run([], _) ->
    lfe_io:format("lfescript: Missing filename\n", []),
    halt(?ERROR_STATUS).

%% parse_check_run(FileName, Args, Options) -> no_return().
%%  Parse the script file, check the code, build a function
%%  environment and call the main function. This function never checks
%%  return values, it just assumes that when a script error is
%%  detected/occurs the functions will call halt(?ERROR_STATUS).
%%  Non-handled errors are program errors.

parse_check_run(File, Args, Lopts) ->
    Fs0 = parse_file(File, Args, Lopts),
    {Fs1,Fenv0} = expand_macros(Fs0, File, Args, Lopts),
    check_code(Fs1, File, Args, Lopts),
    lists:member("s", Lopts) andalso halt(?OK_STATUS),
    Fenv1 = make_env(Fs1, Fenv0, File, Args, Lopts),
    eval_code(Fenv1, File, Args, Lopts),
    halt(?OK_STATUS).                %Everything worked, just exit

error_exit(File, Es, Ws) ->
    list_errors(File, Es),
    list_warnings(File, Ws),
    halt(?ERROR_STATUS).

list_warnings(File, [{Line,Mod,Error}|Ws]) ->
    Cs = Mod:format_error(Error),
    lfe_io:format("~s:~w: Warning: ~s\n", [File,Line,Cs]),
    list_warnings(File, Ws);
list_warnings(_, []) -> ok.

list_errors(File, [{Line,Mod,Error}|Es]) ->
    Cs = Mod:format_error(Error),
    lfe_io:format("~s:~w: ~s\n", [File,Line,Cs]),
    list_errors(File, Es);
list_errors(_, []) -> ok.

%% parse_file(FileName, Args, Lopts) -> [{Expr,Line}].
%% parse_file(FileName) -> {ok,[{Expr,Line}]} | {error,Error}.
%%  This code is copied from lfe_io. We need to do that as we special
%%  case ignoring the first line.

parse_file(File, _, _) ->
    case parse_file(File) of
        {ok,Fs} -> Fs;
        {error,Error} ->
            error_exit(File, [Error], [])
    end.

parse_file(File) ->
    case file:open(File, [read]) of
        {ok,F} ->
            io:get_line(F, ''),                 %Skip first line
            case io:request(F, {get_until,unicode,'',lfe_scan,tokens,[2]}) of
                {ok,Ts,_} ->
                    Ret = parse_file1(Ts, [], []),
                    file:close(F),
                    Ret;
                {error,Error,_} -> {error,Error}
            end;
        {error,Error} -> {error,{none,file,Error}}
    end.

parse_file1([_|_]=Ts0, Pc0, Ss) ->
    case lfe_parse:sexpr(Pc0, Ts0) of
        {ok,L,S,Ts1} -> parse_file1(Ts1, [], [{S,L}|Ss]);
        {more,Pc1} ->
            %% Need more tokens but there are none, so call again to
            %% generate an error message.
            {error,E,_} = lfe_parse:sexpr(Pc1, {eof,99999}),
            {error,E};
        {error,E,_} -> {error,E}
    end;
parse_file1([], _, Ss) -> {ok,lists:reverse(Ss)}.

%% expand_macros(Forms, File, Args, Lopts) -> {Forms,Fenv}.

expand_macros(Fs0, File, _, _) ->
    case lfe_macro:expand_fileforms(Fs0, lfe_env:new(), true, false) of
        {ok,Fs1,Fenv,Ws} ->
            list_warnings(File, Ws),
            {Fs1,Fenv};
        {error,Es,Ws} -> error_exit(File, Es, Ws)
    end.

%% check_code(Forms, File, Args, Lopts) -> ok.
%%  Call lfe_lint to check the code. Must create a dummy module to
%%  make lfe_lint happy.

check_code(Fs, File, _, _) ->
    Module = [{['define-module',dummy,[],[[export,[main,1]]]],1}|Fs],
    case lfe_lint:module(Module) of
        {ok,dummy,Ws} ->
            list_warnings(File, Ws);
        {error,Es,Ws} -> error_exit(File, Es, Ws)
    end.

%% make_env(Forms, File, Args, Lopts) -> FunctionEnv.

make_env(Forms, Fenv, _, _, _) ->
    {Fbs,null} = lfe_lib:proc_forms(fun collect_function/3, Forms, null),
    lfe_eval:make_letrec_env(Fbs, Fenv).

collect_function(['define-function',F,_Meta,Def], _, St) ->
    Ar = function_arity(Def),
    {[{F,Ar,Def}],St};
%% Ignore everything else including types and eval-when-compile.
collect_function(_Form, _, St) ->
    {[],St}.


function_arity([lambda,As|_]) -> length(As);
function_arity(['match-lambda',[Pats|_]|_]) -> length(Pats).


%% eval_code(Fenv, File, Args, Lopts) -> Res.
%%  Evaluate the code. We must explicitly catch and handle errors in
%%  the script.

eval_code(Fenv, _, Args, _) ->
    try
        lfe_eval:expr([main,[quote,Args]], Fenv)
    catch
        %% Catch all exceptions in the code.
        ?CATCH(Class, Error, Stack)
            Skip = fun (_) -> false end,
            Format = fun (T, I) -> lfe_io:prettyprint1(T, 15, I, 80) end,
            Cs = lfe_lib:format_exception(Class, Error, Stack, Skip, Format, 1),
            io:put_chars(Cs),
            halt(?ERROR_STATUS)
    end.
